### ISSUE TITLE
Support multiple sets of credentials

### ISSUE THREAD

**Issue by nathany** [marcelog/elixir_authorizenet#3]
I'm working on a multi-tenant app where companies will be able to select a payment processor. Each company would have different credentials for Authorize.net.

Currently elixir_authorizenet only supports credentials via config.
https://github.com/marcelog/elixir_authorizenet/blob/master/lib/elixir_authorizenet.ex

I'm wondering if there is an elegant way to allow credentials to be specified with more granularity (perhaps per request) while still supporting the config-based behaviour for most users?

**Comment by marcelog** [comment]
Hello!

Not as it is, although I see your point. 

Passing a "config" object in each request "is doable" (and I would deprecate the current behavior of getting it from the application config).

I think I would prefer this over having something like "profiles" setup in the config (each one with its own options and credentials) so they can be dynamically created.

What are your thoughts? Would you be able to work on a patch? If not, I'm willing to, but I need to find some time to do it :)

Best,

**Comment by nathany** [comment]
Hi Marcelo,

I'd be willing to work on a patch once I have some direction. My first thought was "make Gateway an object" and then I was like "oh, right, no objects." (I've only been using Elixir for 4 months).

In our case, we would need to grab the credentials from a database, not from the config. The number of "profiles" could change while the app is running. Of course we'd still want it to be easy for people doing the more typical config-based credentials thing without specifying parameters everywhere.

Thanks.

**Comment by marcelog** [comment]
Hi @nathany,

I agree, although there will be an additional argument needed in each call so the right credentials can be used (you can send a profile id or the full credentials, but something will still have to be there :\ ).

One could send "nil" to specify the current behavior, although I think I'd prefer to just make this new parameter mandatory (i.e: nil not allowed) and just make it work always in the same way (the caller would be responsible for putting the right values there and we would deprecate the config option).

Basically every function calling Main.req will need this new argument, that could be a map that contains the values currently setup in the config section (environment, validation_mode, login_id, and transaction_key, an additional "uri" argument would help with the tests as they are right now, this is due to how the target uri is automatically chosen based on the environment).

Thoughts?

**Comment by nathany** [comment]
That would work for me. I may look at some other payment libraries for Elixir to see if/how they approach this.

**Comment by nathany** [comment]
@marcelog It looks like the client I was doing this for may not use Authorize.net, so I won't have a chance to work on this. I'll let you know if they decide to use it.

**Comment by marcelog** [comment]
No worries 👍 Thanks for letting me know!

Best!
